spark-bigtable_2.12/src/main/scala/com/google/cloud/spark/bigtable/BigtableDefaultSource.scala

```diff
@@ -0,0 +1,200 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spark.bigtable
+
+import com.google.cloud.bigtable.admin.v2.BigtableTableAdminClient
+import com.google.cloud.bigtable.admin.v2.models.CreateTableRequest
+import com.google.cloud.spark.bigtable.datasources._
+import com.google.cloud.spark.bigtable.filters.{
+  RowKeyWrapper,
+  SparkSqlFilterAdapter
+}
+import com.google.common.collect.RangeSet
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{DataFrame, SQLContext, SaveMode, Row => SparkRow}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.yetus.audience.InterfaceAudience
+
+object VersionInformation {
+  val CONNECTOR_VERSION = "0.1.0"
+  val DATA_SOURCE_VERSION = "V1"
+  val scalaVersion = util.Properties.versionNumberString
+  // This remains unset only in unit tests where sqlContext is null.
+  var sparkVersion = "UNSET_SPARK_VERSION"
+}
+
+/** Bigtable DefaultSource class which creates a BigtableRelation object.
+  */
+@InterfaceAudience.Private
+class BigtableDefaultSource
+    extends RelationProvider
+    with CreatableRelationProvider
+    with DataSourceRegister {
+
+  /** Constructs a BigtableRelation object.
+    *
+    * @param sqlContext Spark SQL context
+    * @param parameters Parameters from Spark SQL
+    * @return           A BigtableRelation object
+    */
+  override def createRelation(
+      sqlContext: SQLContext,
+      parameters: Map[String, String]
+  ): BaseRelation = {
+    new BigtableRelation(parameters, None)(sqlContext)
+  }
+
+  /** Constructs a BigtableRelation object after writing the provided data
+    *   to Bigtable.
+    *
+    * @param sqlContext Spark SQL context
+    * @param mode       The save mode
+    * @param parameters Parameters from Spark SQL
+    * @param data       The data to write to Bigtable
+    * @return           A BigtableRelation object
+    */
+  override def createRelation(
+      sqlContext: SQLContext,
+      mode: SaveMode,
+      parameters: Map[String, String],
+      data: DataFrame
+  ): BaseRelation = {
+    val relation = BigtableRelation(parameters, Some(data.schema))(sqlContext)
+    relation.createTableIfNeeded()
+    relation.insert(data, false)
+    relation
+  }
+
+  override def shortName(): String = "bigtable"
+}
+
+/** Custom Relation class which manages reading from and writing to Bigtable.
+  *
+  * @param sqlContext              Spark SQL context
+  */
+@InterfaceAudience.Private
+case class BigtableRelation(
+    @transient parameters: Map[String, String],
+    userSpecifiedSchema: Option[StructType]
+)(@transient val sqlContext: SQLContext)
+    extends BaseRelation
+    with PrunedFilteredScan
+    with InsertableRelation
+    with Logging {
+  Option(sqlContext).foreach(context =>
+    VersionInformation.sparkVersion = context.sparkContext.version
+  )
+  val pushDownRowKeyFilters: Boolean = parameters
```

Comment:
Since these variables remain the same after each instantiation of the relation, I think we can keep them here since they'll be accessed in different places (sometimes when reading, sometimes in RDDs, etc.)